chore(docs): update docs (#1363)
Satellite-im/Uplink: docs/architecture.md

##  Uplink Architecture
- This document should serve as a high-level guide to the Uplink architecture and explain various ways to add to the application
---

## Background
- The Uplink repository consists of two `Cargo` projects: `kit` (a library) and `uplink` (the executable).  
- Uplink relies on [Warp](https://github.com/Satellite-im/Warp) and [Dioxus](https://github.com/DioxusLabs/dioxus). It is assumed the reader is familiar with the [Dioxus Documentation](https://dioxuslabs.com/guide/). 

> - At a high level, Warp sends messages (it does much more), while Uplink is just the UI. Sending and receiving messages is asynchronous, and Uplink has a separate module to handle this: `warp_runner`. All the data for the Uplink UI is contained in a `State` struct. Changing the `State` struct drives the UI. When `warp_runner` handles an event from `Warp`, `State` is modified, and the entire UI is re-rendered.   

Author: ```suggestion
- At a high level, Warp sends messages (it does much more), while Uplink is just the UI. Sending and receiving messages is asynchronous, and Uplink has a separate module to handle this: `warp_runner`. All the data for the Uplink UI is contained in a `State` struct. Changing the `State` struct drives the UI. When `warp_runner` handles an event from the `Warp`, `State` is modified, and the entire UI is re-rendered.   
```

- use `logger::trace` to track when elements render
- get a hook for `State` via `use_shared_state`—no need to pass it in Props. 

## Starting the UI
- in `main.rs`, there is a call to `dioxus_desktop::launch_cfg`. It is passed a `bootstrap` element, which is rendered first. 
- Uplink needs to start `warp_runner` from within a `Tokio` runtime. That runtime isn't available until `dioxus_desktop::launch_cfg` is called. We don't want to restart `warp_runner` when the UI updates, so the `bootstrap` Element ensures this only happens once. 
- next `auth_page_manager` is rendered. `Warp` requires a user account. Until an account is created, the user data can't be decrypted and loaded. `auth_page_manager` uses the `auth_wrapper` element to handle user authentication and account creation. When done, `app_bootstrap` is rendered. 
- The global `State` variable drives the Uplink UI. It is initialized via `use_shared_state_provider`. We don't want this variable to be re-loaded every time the app updates, so `app_bootstrap` is used to initialize state. 
- finally, `app` is rendered. All the global channels are polled here, each with their own `use_future`. 

>     + conversations are loaded from the warp and added to `State`

Reviewer: ~~the~~ warp

Author: done

>     + conversations are loaded from the warp and added to `State`

Author: ```suggestion
    + conversations are loaded from warp and added to `State`
```

>     + friends are loaded from the warp and added to `State`

Reviewer: ~~the~~ warp

Author: done

>     + friends are loaded from the warp and added to `State`

Author: ```suggestion
    + friends are loaded from warp and added to `State`
```

    + warp events are passed to `State::process_warp_event`
    + after all the `use_future`s, `cx.render()` is called with the various parts of the page.

## Global Variables
- Uplink uses global variables to do some essential things. They are all in `main.rs`. 

> - `STATIC_ARGS` is for constant variables that may be affected by command line options

Reviewer: like 'which' better here

Author: done

> - `STATIC_ARGS` is for constant variables that may be affected by command line options

Author: ```suggestion
- `STATIC_ARGS` is for constant variables which may be affected by command line options
```

- the rest of the variables are channels. 

## Global Channels

> - `WARP_CMD_CH`: used to communicate with `warp_runner`. Rather than block the UI while `Warp` performs a task, `warp_runner` receives commands via the channel and responds via a one-shot channel. `WARP_CMD_CH.tx` is used from within a Dioxus Element, inside a `use_future` or a `use_coroutine`. 

Reviewer: "oneshot" is the actual name of something

Author: done

> - `WARP_CMD_CH`: used to communicate with `warp_runner`. Rather than block the UI while `Warp` performs a task, `warp_runner` receives commands via the channel and responds via a one-shot channel. `WARP_CMD_CH.tx` is used from within a Dioxus Element, inside a `use_future` or a `use_coroutine`. 

Author: ```suggestion
- `WARP_CMD_CH`: used to communicate with `warp_runner`. Rather than block the UI while `Warp` performs a task, `warp_runner` receives commands via the channel and responds via a oneshot channel. `WARP_CMD_CH.tx` is used from within a Dioxus Element, inside a `use_future` or a `use_coroutine`. 
```

- `WARP_EVENT_CH`: the `app` Element from `main.rs` reads events from this channel and updates `State` accordingly. 
- `WINDOW_CMD_CH`: if Uplink spawns a child window, it is controlled via this channel. Commands are passed to the `window_manager` module, which, at a minimum, allows for closing the window in response to a UI event such as a button press. 

## warp_runner::WarpRunner

> - provides access to `RayGun`, `MultiPass`, and `Tesseract` via the `WarpCmd` struct. Commands contain a one-shot channel for returning the result. `WARP_CMD_CH` is used to send `WarpCmd`s to `WarpRunner`.   

Reviewer: "oneshot"

Author: done

> - provides access to `RayGun`, `MultiPass`, and `Tesseract` via the `WarpCmd` struct. Commands contain a one-shot channel for returning the result. `WARP_CMD_CH` is used to send `WarpCmd`s to `WarpRunner`.   

Author: ```suggestion
- provides access to `RayGun`, `MultiPass`, and `Tesseract` via the `WarpCmd` struct. Commands contain a oneshot channel for returning the result. `WARP_CMD_CH` is used to send `WarpCmd`s to `WarpRunner`.   
```
